signature/src/policy.rs

```diff
@@ -0,0 +1,824 @@
+        // Get the policy set that matches the image.
+        let reqs = self.requirements_for_image(&image);
+        if reqs.is_empty() {
+            return Err(anyhow!(
+                "List of verification policy requirements must not be empty"
+            ));
+        }
+
+        // The image must meet the requirements of each policy in the policy set.
+        for req in reqs.iter() {
+            req.is_running_image_allowed(&mut image)?;
+        }
+
+        Ok(())
+    }
+
+    // selects the appropriate requirements for the image from Policy.
+    fn requirements_for_image(&self, image: &image::Image) -> PolicyRequirements {
+        // Get transport name of the image,
+        // FIXME: now only support "docker" (image from the registry which implement Docker HTTPS APIs V2)
+        let transport_name = image.transport_name();
+
+        if let Some(transport_scopes) = self.transports.get(&transport_name) {
+            // Look for a full match.
+            let identity = image::get_image_identity(&image.reference);
+            if let Some(reqs) = transport_scopes.get(&identity) {
+                return reqs.clone();
+            }
+
+            // Look for a match of the possible parent namespaces.
+            for name in image::get_image_namespace(&image.reference).iter() {
+                if let Some(reqs) = transport_scopes.get(name) {
+                    return reqs.clone();
+                }
+            }
+
+            // Look for a default match for the transport.
+            if let Some(reqs) = transport_scopes.get("") {
+                return reqs.clone();
+            }
+        }
+
+        self.default.clone()
+    }
+}
+
+// PolicyRequirement is a rule which must be satisfied.
+// It has three types:
+//   `insecureAcceptAnything`: skip signature verification, accept the image unconditionally.
+//   `reject`: reject the image directly.
+//   `signedBy`: there must be at least a signature of the image can be verified by the specific key.
+// The first two are represented by PrSimple, the last is represented by PrSignedBy.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+#[serde(untagged)]
+pub enum PolicyRequirement {
+    Signedby(PrSignedBy),
+    Simple(PrSimple),
+}
+
+impl TryFrom<&str> for PolicyRequirement {
+    type Error = serde_json::Error;
+    fn try_from(json_str: &str) -> Result<Self, Self::Error> {
+        serde_json::from_str::<PolicyRequirement>(json_str)
+    }
+}
+
+impl PolicyRequirement {
+    pub fn is_running_image_allowed(&self, image: &mut image::Image) -> Result<()> {
+        match self {
+            PolicyRequirement::Simple(req) => req.simple_judge_allowed(image),
+            PolicyRequirement::Signedby(req) => req.signature_judge_allowed(image),
+        }
+    }
+}
+
+// The `insecureAcceptAnything` or `reject` policy requirement.
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+pub struct PrSimple {
```

Comment:
For me, It's very hard to see the letters "pr" and not think "pull request" :smile: But this PR seems to overload the  acronym further since fwics it can mean:

- "policy requirement"
- "policy reference"

I'd strongly consider renaming some of the types to make this distinction clearer and make the code less confusing:

`PolicyReqSimple`, `PolicyReqSignedBy`, `PolicyRefMatch`, etc_?